src/transport/SecurePairingSession.cpp

```diff
@@ -218,68 +280,242 @@ CHIP_ERROR SecurePairingSession::AttachHeaderAndSend(uint8_t msgType, System::Pa
+    err = mCommissioningHash.AddData(req->Start(), req->TotalLength());
+    SuccessOrExit(err);
+
+    mNextExpectedMsg = Spake2pMsgType::kPBKDFParamResponse;
+
+    err = AttachHeaderAndSend(Spake2pMsgType::kPBKDFParamRequest, req.Release_ForNow());
+    SuccessOrExit(err);
+
+    ChipLogDetail(Ble, "Sent PBKDF param request");
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        Clear();
+    }
+    return err;
+}
+
+CHIP_ERROR SecurePairingSession::HandlePBKDFParamRequest(const PacketHeader & header, const System::PacketBufferHandle & msg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Request message processing
+    const uint8_t * req = msg->Start();
+    size_t reqlen       = msg->TotalLength();
+
+    VerifyOrExit(req != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(reqlen == sizeof(PBKDFParamRequest), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    ChipLogDetail(Ble, "Received PBKDF param request");
+
+    // Update commissioning hash with the received pbkdf2 param request
+    err = mCommissioningHash.AddData(req, reqlen);
+    SuccessOrExit(err);
+
+    err = SendPBKDFParamResponse();
+    SuccessOrExit(err);
+
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(Spake2pErrorType::kUnexpected);
+    }
+    return err;
+}
+
+CHIP_ERROR SecurePairingSession::SendPBKDFParamResponse()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    System::PacketBufferHandle resp;
+    uint16_t resplen           = 0;
+    PBKDFParamResponse * pResp = nullptr;
+
+    size_t sizeof_point = sizeof(mPoint);
+
+    VerifyOrExit(CanCastTo<uint16_t>(sizeof(PBKDFParamResponse) + mSaltLength), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    resplen = static_cast<uint16_t>(sizeof(PBKDFParamResponse) + mSaltLength);
+
+    resp = System::PacketBuffer::NewWithAvailableSize(resplen);
     VerifyOrExit(!resp.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
 
+    pResp = reinterpret_cast<PBKDFParamResponse *>(resp->Start());
+    err   = DRBG_get_bytes(pResp->mRandom, sizeof(pResp->mRandom));
+    SuccessOrExit(err);
+
+    chip::Encoding::LittleEndian::Put64(reinterpret_cast<uint8_t *>(&pResp->mIterations), mIterationCount);
+    chip::Encoding::LittleEndian::Put32(reinterpret_cast<uint8_t *>(&pResp->mSaltLength), mSaltLength);
+
     {
-        BufBound bbuf(resp->Start(), data_len);
-        bbuf.Put(&X[0], X_len);
+        BufBound bbuf(pResp->mSalt, mSaltLength);
+        bbuf.Put(mSalt, mSaltLength);
         VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
     }
 
-    resp->SetDataLength(data_len);
-    mNextExpectedMsg = Spake2pMsgType::kSpake2pCompute_pB_cB;
+    resp->SetDataLength(resplen);
 
-    // Call delegate to send the Compute_pA to peer
-    err = AttachHeaderAndSend(Spake2pMsgType::kSpake2pCompute_pA, resp.Release_ForNow());
+    // Update commissioning hash with the pbkdf2 param response that's being sent.
+    err = mCommissioningHash.AddData(resp->Start(), resp->TotalLength());
```

Comment:
If there is any sort of compiler-inserted padding in the struct, due to alignment requirements, we will have some random bytes that we include in the hash here.  And if the other side was compiled with a different compiler that has different padding going on, things won't match up properly.

More generally, these overlay structs seem to me to be more trouble than they're worth.  If we were using them for something other than this encoding/decoding bit, they'd make some sense, but for this use they seem to be more trouble than they're worth, given the various freedom C/C++ compilers have in terms of struct layout.

We might be able to get away with some static asserts somewhere about `sizeof(PBKDFParamResponse)` is exactly `kPBKDFParamRandomNumberSize + 8 + 4` or something like that, but it might be simpler to just use a BufBound on the writer end that skips `kPBKDFParamRandomNumberSize` bytes that we do `DRBG_get_bytes` into, then writes out the iteration count and salt length, then writes out the salt.  And on the reader end we just read the data out, without a struct being involved.